File: Graph Theory Peaks/Union Find/Q2. Largest Component Size by Common Factor/Q2_LargestComponentSizebyCommonFactor.java
class Solution {

    // Stores the parent of each node (Disjoint Set / Union-Find)
    HashMap<Integer, Integer> parent;

    // Stores the size of each connected component (only valid for root nodes)
    HashMap<Integer, Integer> size;

    // Initialize the parent and size maps
    public void init(){
        parent = new HashMap<>();
        size = new HashMap<>();
    }

    // Create a new group (set) for number n
    public void newGroup(int n){
        // If already exists, do nothing
        if(parent.containsKey(n)) return;

        // Initially, a node is its own parent (self root)
        parent.put(n, n);

        // Initial size of each component is 1
        size.put(n, 1);
    }

    // Find the ultimate parent (root) of x
    public int find(int x){
        // If x is its own parent, it is the root
        if(parent.get(x)==x){
            return x;
        }

        // Recursively find the root
        return find(parent.get(x));
    }

    // Union two nodes a and b (merge their components)
    public void union(int a, int b){

        // Find root of a
        int parentA = find(a);

        // Find root of b
        int parentB = find(b);

        // If both have same parent, already connected
        if(parentA==parentB) return;

        // Union by size (attach smaller tree under bigger tree)
        if(size.get(parentA)>=size.get(parentB)){

            // Make parentB's root point to parentA
            parent.put(parentB, parentA);

            // Increase size of parentA component
            size.put(parentA, size.get(parentA)+size.get(parentB));
        }else{

            // Make parentA's root point to parentB
            parent.put(parentA, parentB);

            // Increase size of parentB component
            size.put(parentB, size.get(parentA)+size.get(parentB));
        }
    }

    // Get all factors (excluding 1) of number a
    public ArrayList<Integer> getFactors(int a){

        // List to store factors
        ArrayList<Integer> list = new ArrayList<>();

        // Iterate from 2 to sqrt(a)
        for(int i=2; i<=Math.sqrt(a); i++){

            // If i divides a, then both i and a/i are factors
            if(a%i==0){
                list.add(i);
                list.add(a/i);
            }
        }

        // Add the number itself (important for prime numbers)
        list.add(a);

        return list;
    }

    public int largestComponentSize(int[] nums) {

        int n = nums.length;

        // Map: factor -> list of numbers having that factor
        // This acts like a graph connection via common factors
        HashMap<Integer, ArrayList<Integer>> map = new HashMap<>();

        // Initialize DSU structure
        init();

        // Step 1: Build groups and factor mapping
        for(int num : nums){

            // Make each number its own set initially
            newGroup(num);

            // Get factors of num
            ArrayList<Integer> factors = new ArrayList<>(getFactors(num));

            // Map each factor to the numbers having that factor
            for(int f : factors){

                // If factor not present in map, create list
                map.putIfAbsent(f, new ArrayList<>());

                // Add current number to factor's list
                ArrayList<Integer> list = map.get(f);
                list.add(num);
            }
        }

        // Step 2: Union numbers that share the same factor
        for(int key : map.keySet()){

            // Get all numbers that share this factor
            ArrayList<Integer> list = map.get(key);

            // Union consecutive numbers in the list
            for(int i=0; i<list.size()-1; i++){

                // Connect list[i] and list[i+1]
                union(list.get(i), list.get(i+1));
            }
        }

        // Step 3: Find maximum component size
        int max = Integer.MIN_VALUE;

        for(int num : nums){

            // Find root of this number
            int par = find(num);

            // Update maximum size among components
            max = Math.max(size.get(par), max);
        }

        // Return largest connected component size
        return max;
    }
}

File: Graph Theory Peaks/Union Find/Q2. Largest Component Size by Common Factor/Q2_LargestComponentSizebyCommonFactor.py
class Solution(object):
    def largestComponentSize(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        import collections

        # Step 1: Generate prime numbers up to sqrt(max(nums))
        # We only need primes up to sqrt(max value) for factorization
        primes = []
        for x in range(2, int(max(nums) ** 0.5) + 1):
            for y in primes:
                # If x is divisible by any existing prime, it's not prime
                if x % y == 0:
                    break
            else:
                # If no divisor found, x is prime
                primes.append(x)

        # Step 2: Compute prime factors for each number
        # factors[a] will store unique prime factors of number a
        factors = collections.defaultdict(list)

        for a in nums:
            x = a  # Copy of number for factorization

            # Try dividing by all primes
            for p in primes:
                # If p^2 > x, no further factorization needed
                if p * p > x:
                    break

                # If p divides x, it's a prime factor
                if x % p == 0:
                    factors[a].append(p)

                    # Remove all occurrences of p from x
                    while x % p == 0:
                        x //= p

            # If remainder x > 1, it is also a prime factor
            if x > 1:
                factors[a].append(x)
                primes.append(x)  # Add new discovered prime

        # Remove duplicate primes
        primes = list(set(primes))
        n = len(primes)

        # Map each prime to an index (for union-find)
        p2i = {p: i for i, p in enumerate(primes)}

        # Step 3: Initialize Union-Find (DSU) on primes
        parent = [i for i in range(n)]

        # Find function with path compression
        def find(i):
            if i != parent[i]:
                parent[i] = find(parent[i])  # Compress path
            return parent[i]

        # Union two prime indices
        def union(i, j):
            pi, pj = find(i), find(j)
            if pi != pj:
                parent[pi] = pj  # Merge sets

        # Step 4: Connect primes that appear in same number
        for a in nums:
            if factors[a]:  # If number has prime factors
                p0 = factors[a][0]  # Take first prime factor

                # Union first prime with all other primes of same number
                for p in factors[a][1:]:
                    union(p2i[p0], p2i[p])

        # Step 5: Count how many numbers belong to each connected prime component
        # Each number corresponds to the root of its first prime factor
        count = collections.Counter(
            find(p2i[factors[a][0]]) for a in nums if factors[a]
        )

        # Return size of largest connected component
        return max(count.values())